doc/source/whatsnew/v0.25.0.rst

```diff
@@ -514,6 +514,7 @@ Performance Improvements
 - Improved performance of :meth:`read_csv` by faster concatenating date columns without extra conversion to string for integer/float zero and float ``NaN``; by faster checking the string for the possibility of being a date (:issue:`25754`)
 - Improved performance of :attr:`IntervalIndex.is_unique` by removing conversion to ``MultiIndex`` (:issue:`24813`)
 - Restored performance of :meth:`DatetimeIndex.__iter__` by re-enabling specialized code path (:issue:`26702`)
+- Improved performance of :meth:`DataFrame.set_index` when using multiple indexes and at least one of them is a categorical object (:issue:`22044`)
```

Comment:
can you add the asv as you have in the example. Also we might not have tests for using multiple cats in a MI, see if you can add one in pandas/tests/indexes/multi/test_constructor which match what the .set_index() is doing at a lower level.